### ISSUE TITLE
bug: Incorrect ScalarMultiplication  in bandersnatch

### ISSUE THREAD

**Issue by liyue201** [Consensys/gnark-crypto#515]
## Description

ScalarMultiplication returns incorrect results when certain parameters are passed in.
This is my code and it returns the correct result when I change the scalar.

```go
package main

import (
	"fmt"
	"github.com/consensys/gnark-crypto/ecc/bls12-381/bandersnatch"
	"github.com/consensys/gnark-crypto/ecc/bls12-381/fr"
	"math/big"
)

var Identity = bandersnatch.PointProj{
	X: fr.Element{},
	Y: fr.One(),
	Z: fr.One(),
}

func scalaMul(point *bandersnatch.PointProj, scalar *big.Int) *bandersnatch.PointProj {
	var res bandersnatch.PointProj
	var d bandersnatch.PointProj
	res.Set(&Identity)
	d.Set(point)

	for i := 0; i < scalar.BitLen(); i++ {
		b := scalar.Bit(i)
		if b == 1 {
			res.Add(&res, &d)
		}
		if i < scalar.BitLen()-1 {
			d.Double(&d)
		}
	}
	return &res
}

func toProj(p *bandersnatch.PointAffine) *bandersnatch.PointProj {
	var pointProject bandersnatch.PointProj
	pointProject.FromAffine(p)
	return &pointProject
}

func main() {
	var affinePoint bandersnatch.PointAffine
	affinePoint.X.SetString("18117916180900044038769101307378694226656604001413977957989208801045106289272")
	affinePoint.Y.SetString("32622399107551617669424155920349902745880508009388682895260105611357660813203")

	projPoint := toProj(&affinePoint)

	scalar, _ := new(big.Int).SetString("175765906040938024221791434587452274757738266628326740483090387010754277314", 10)
	//scalar, _ := new(big.Int).SetString("6475475475", 10)

	res1 := scalaMul(projPoint, scalar)
	res2 := new(bandersnatch.PointProj).ScalarMultiplication(projPoint, scalar)

	fmt.Printf("res1: %v\n", res1)
	fmt.Printf("res2: %v\n", res2)

	if !res1.Equal(res2) {
		panic("ScalarMultiplication error")
	}
}

```

## Steps to Reproduce

1.
2.
3.
4.

## Your Environment

* gnark-crypto version used:
* go version (e.g. 1.20.6):
* Operating System and version:

**Comment by yelhousni** [comment]
I think the issue is the input point and not the scalar. This method
```go
res2 := new(bandersnatch.PointProj).ScalarMultiplication(projPoint, scalar)
```
uses `scalarMulGMV` underneath
https://github.com/Consensys/gnark-crypto/blob/1ef1276c13345e9a8693470f61913be26e444029/ecc/bls12-381/bandersnatch/endomorpism.go#L35
which assumes that the point is of order 
https://github.com/Consensys/gnark-crypto/blob/1ef1276c13345e9a8693470f61913be26e444029/ecc/bls12-381/bandersnatch/curve.go#L67
If not the endomorphism is not well defined and the result is incorrect. You need to clear the cofactor with 
```go
// multiply by the cofactor 4
affinePoint.Double(&affinePoint).Double(&affinePoint)
```
and then the `res1` and `res2` are equal.

**Comment by yelhousni** [comment]
This is said there is a (potential) _distinct_ bug with `ScalarMultiplication ` for Bandersnatch. The loop bounds in `scalarMulGLV` should conditionally be increased by 1. After we revisited GLV scalar decomposition in https://github.com/Consensys/gnark-crypto/pull/213, we did so for all Weierstrass curves but forgot about twisted Edwards (i.e. Bandersnatch). The PR https://github.com/Consensys/gnark-crypto/pull/213 fixes this.
